File: lua/mpp/protocol/solana.lua
local M = {
  TOKEN_PROGRAM = 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
  TOKEN_2022_PROGRAM = 'TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb',
  ASSOCIATED_TOKEN_PROGRAM = 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL',
  SYSTEM_PROGRAM = '11111111111111111111111111111111',
  MEMO_PROGRAM = 'MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr',
}

local KNOWN_MINTS = {
  USDC = {
    devnet = '4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU',
    testnet = '4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU',
    ['mainnet-beta'] = 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v',
  },
  USDT = {
    ['mainnet-beta'] = 'Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB',
  },
  USDG = {
    devnet = '4F6PM96JJxngmHnZLBh9n58RH4aTVNWvDs2nuwrT5BP7',
    testnet = '4F6PM96JJxngmHnZLBh9n58RH4aTVNWvDs2nuwrT5BP7',
    ['mainnet-beta'] = '2u1tszSeqZ3qBWF3uNGPFc8TzMk2tdiwknnRMWGWjGWH',
  },
  PYUSD = {
    devnet = 'CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM',
    testnet = 'CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM',
    ['mainnet-beta'] = '2b1kV6DkPAnxd5ixfnxCpjxmKwqjjaYmCZfHsFu24GXo',
  },
  CASH = {
    ['mainnet-beta'] = 'CASHx9KJUStyftLFWGvEVf59SGeG9sh5FfcnZMVPCASH',
  },
}

local TOKEN_PROGRAMS = {
  USDC = M.TOKEN_PROGRAM,
  USDT = M.TOKEN_PROGRAM,
  USDG = M.TOKEN_2022_PROGRAM,
  PYUSD = M.TOKEN_2022_PROGRAM,
  CASH = M.TOKEN_2022_PROGRAM,
}

function M.default_rpc_url(network)
  if network == 'devnet' then
    return 'https://api.devnet.solana.com'
  elseif network == 'localnet' then
    return 'http://localhost:8899'
  end
  return 'https://api.mainnet-beta.solana.com'
end

function M.resolve_mint(currency, network)
  local normalized = string.upper(currency or '')
  if normalized == 'SOL' then
    return nil
  end

  local known = KNOWN_MINTS[normalized]
  if known then
    return known[network] or known['mainnet-beta']
  end

  return currency
end

function M.stablecoin_symbol(currency)
  local normalized = string.upper(currency or '')
  if KNOWN_MINTS[normalized] then
    return normalized
  end
  for symbol, mints in pairs(KNOWN_MINTS) do
    for _, mint in pairs(mints) do
      if currency == mint then
        return symbol
      end
    end
  end
  return nil
end

function M.default_token_program_for_currency(currency, network)
  local symbol = M.stablecoin_symbol(M.resolve_mint(currency, network)) or M.stablecoin_symbol(currency)
  if symbol then
    return TOKEN_PROGRAMS[symbol]
  end
  return M.TOKEN_PROGRAM
end

return M

File: go/protocol/solana.go
package protocol

import "strings"

const (
	TokenProgram           = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"
	Token2022Program       = "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb"
	AssociatedTokenProgram = "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL"
	SystemProgram          = "11111111111111111111111111111111"

	USDCMainnetMint  = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
	USDCDevnetMint   = "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU"
	USDTMainnetMint  = "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"
	USDGMainnetMint  = "2u1tszSeqZ3qBWF3uNGPFc8TzMk2tdiwknnRMWGWjGWH"
	USDGDevnetMint   = "4F6PM96JJxngmHnZLBh9n58RH4aTVNWvDs2nuwrT5BP7"
	PYUSDMainnetMint = "2b1kV6DkPAnxd5ixfnxCpjxmKwqjjaYmCZfHsFu24GXo"
	PYUSDDevnetMint  = "CXk2AMBfi3TwaEL2468s6zP8xq9NxTXjp9gjMgzeUynM"
	CASHMainnetMint  = "CASHx9KJUStyftLFWGvEVf59SGeG9sh5FfcnZMVPCASH"
)

var knownMints = map[string]map[string]string{
	"USDC": {
		"devnet":       USDCDevnetMint,
		"testnet":      USDCDevnetMint,
		"mainnet-beta": USDCMainnetMint,
	},
	"USDT": {
		"mainnet-beta": USDTMainnetMint,
	},
	"USDG": {
		"devnet":       USDGDevnetMint,
		"testnet":      USDGDevnetMint,
		"mainnet-beta": USDGMainnetMint,
	},
	"PYUSD": {
		"devnet":       PYUSDDevnetMint,
		"testnet":      PYUSDDevnetMint,
		"mainnet-beta": PYUSDMainnetMint,
	},
	"CASH": {
		"mainnet-beta": CASHMainnetMint,
	},
}

var token2022Stablecoins = map[string]struct{}{
	"PYUSD": {},
	"USDG":  {},
	"CASH":  {},
}

// DefaultRPCURL returns the default RPC endpoint for a Solana network.
func DefaultRPCURL(network string) string {
	switch network {
	case "devnet":
		return "https://api.devnet.solana.com"
	case "localnet":
		return "http://localhost:8899"
	default:
		return "https://api.mainnet-beta.solana.com"
	}
}

// ResolveMint converts a symbolic currency into a mint address.
// Returns an empty string for native SOL.
func ResolveMint(currency string, network string) string {
	normalized := strings.ToUpper(currency)
	switch normalized {
	case "SOL":
		return ""
	}
	if mints, ok := knownMints[normalized]; ok {
		if mint, ok := mints[network]; ok {
			return mint
		}
		return mints["mainnet-beta"]
	}
	return currency
}

// StablecoinSymbol returns a supported stablecoin symbol for a symbol or known mint.
func StablecoinSymbol(currency string) string {
	normalized := strings.ToUpper(currency)
	if _, ok := knownMints[normalized]; ok {
		return normalized
	}
	for symbol, mints := range knownMints {
		for _, mint := range mints {
			if currency == mint {
				return symbol
			}
		}
	}
	return ""
}

// DefaultTokenProgramForCurrency returns the known default token program for a currency or mint.
func DefaultTokenProgramForCurrency(currency string, network string) string {
	symbol := StablecoinSymbol(ResolveMint(currency, network))
	if symbol == "" {
		symbol = StablecoinSymbol(currency)
	}
	if _, ok := token2022Stablecoins[symbol]; ok {
		return Token2022Program
	}
	return TokenProgram
}

// MethodDetails contains Solana-specific challenge fields.
type MethodDetails struct {
	Network         string  `json:"network,omitempty"`
	Decimals        *uint8  `json:"decimals,omitempty"`
	TokenProgram    string  `json:"tokenProgram,omitempty"`
	FeePayer        *bool   `json:"feePayer,omitempty"`
	FeePayerKey     string  `json:"feePayerKey,omitempty"`
	Splits          []Split `json:"splits,omitempty"`
	RecentBlockhash string  `json:"recentBlockhash,omitempty"`
}

// Split is an additional transfer in the same asset.
type Split struct {
	Recipient string `json:"recipient"`
	Amount    string `json:"amount"`
	Label     string `json:"label,omitempty"`
	Memo      string `json:"memo,omitempty"`
}

// CredentialPayload is sent by clients in the payment credential payload.
type CredentialPayload struct {
	Type        string `json:"type"`
	Transaction string `json:"transaction,omitempty"`
	Signature   string `json:"signature,omitempty"`
}
